src/mscorlib/src/System/TimeZoneInfo.cs

```diff
@@ -1944,10 +1971,173 @@ private class OffsetAndRule {
+                    if (tzVariable[0] != '/')
+                    {
+                        id = tzVariable;
+                        tzFilePath = Path.Combine(c_timeZoneDirectory, tzVariable);
+                    }
+                    else
+                    {
+                        tzFilePath = tzVariable;
+                    }
+
+                    if (!File.Exists(tzFilePath))
+                    {
+                        rawData = null;
+                        id = null;
+                        return false;
+                    }
+                }
+            }
+            else
+            {
+                const string localtimeFilePath = "/etc/localtime";
+                if (File.Exists(localtimeFilePath))
+                {
+                    // UNIXTODO: optimzation - see if localtimeFilePath is a symlink, and use 'readlink' to get the id
+                    tzFilePath = localtimeFilePath;
+                }
+
+                if (tzFilePath == null)
+                {
+                    string usrLocalTimeFilePath = Path.Combine(c_timeZoneDirectory, "localtime");
+                    if (File.Exists(usrLocalTimeFilePath))
+                    {
+                        tzFilePath = usrLocalTimeFilePath;
+                    }
+                }
+            }
+
+            if (string.IsNullOrEmpty(tzFilePath))
+            {
+                rawData = null;
+                id = null;
+                return false;
+            }
+
+            try
+            {
+                rawData = File.ReadAllBytes(tzFilePath);
+                if (rawData != null && string.IsNullOrEmpty(id))
+                {
+                    id = FindTimeZoneId(rawData);
+                }
+
+                return rawData != null;
+            }
+            catch (IOException)
+            {
+                rawData = null;
+                id = null;
+                return false;
+            }
+        }
+
+        private static string GetTzEnvironmentVariable()
+        {
+            string result = Environment.GetEnvironmentVariable("TZ");
+            if (!string.IsNullOrEmpty(result))
+            {
+                if (result[0] == ':')
+                {
+                    // strip off the ':' prefix
+                    result = result.Substring(1);
+                }
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// Find the time zone id by searching all the tzfiles for the one that matches rawData
+        /// and return its file name.
+        /// </summary>
+        private static string FindTimeZoneId(byte[] rawData)
+        {
+            string id = null;
+            string localtimeFilePath = Path.Combine(c_timeZoneDirectory, "localtime");
+
+            foreach (string filePath in Directory.EnumerateFiles(c_timeZoneDirectory, "*", SearchOption.AllDirectories))
```

Comment:
The body of the loop is catching IOExceptions for the case where the file can't be opened.  What about if the directory can't be enumerated?  And are there cases where other kinds of exceptions might be thrown, like UnauthorizedAccessException?